src/test/java/org/junitpioneer/jupiter/displaynamegenerator/ReplaceCamelCaseAndUnderscoreAndNumberTestEngine.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v2.0 which
+ * accompanies this distribution and is available at
+ *
+ * http://www.eclipse.org/legal/epl-v20.html
+ */
+
+package org.junitpioneer.jupiter.displaynamegenerator;
+
+import java.util.Optional;
+
+import org.junit.platform.engine.EngineDiscoveryRequest;
+import org.junit.platform.engine.ExecutionRequest;
+import org.junit.platform.engine.TestDescriptor;
+import org.junit.platform.engine.TestEngine;
+import org.junit.platform.engine.UniqueId;
+import org.junit.platform.engine.discovery.ClassSelector;
+import org.junit.platform.engine.support.descriptor.EngineDescriptor;
+import org.junit.platform.launcher.LauncherDiscoveryRequest;
+
+public class ReplaceCamelCaseAndUnderscoreAndNumberTestEngine implements TestEngine {
+
+	@Override
+	public String getId() {
+		return "replace-camelcase-and-underscore-and-number";
+	}
+
+	@Override
+	public TestDescriptor discover(EngineDiscoveryRequest engineDiscoveryRequest, UniqueId uniqueId) {
+		TestDescriptor engineDescriptor = new EngineDescriptor(uniqueId, "ReplaceCamelCaseAndUnderscoreAndNumber Test");
+		engineDiscoveryRequest.getSelectorsByType(ClassSelector.class).forEach(selector -> {
+			appendTestsInClass(selector.getJavaClass(), engineDescriptor);
+		});
+		return engineDescriptor;
+	}
+
+	private void appendTestsInClass(Class<?> javaClass, TestDescriptor engineDescriptor) {
+		engineDescriptor.addChild(new ClassTestDescriptor(javaClass, engineDescriptor));
+	}
+
+	@Override
+	public void execute(ExecutionRequest executionRequest) {
+		TestDescriptor rootTestDescriptor = executionRequest.getRootTestDescriptor();
+		TestExecutor testExecutor = new TestExecutor(executionRequest.getEngineExecutionListener());
+		testExecutor.execute(rootTestDescriptor);
+
+	}
+
+	@Override
+	public Optional<String> getGroupId() {
+		return Optional.of("org.junitpioneer.jupiter");
+	}
+
+	@Override
+	public Optional<String> getArtifactId() {
+		return Optional.of("replace-camelcase-and-underscore-and-number-engine");
```

Comment:
I'm not familiar with `TestEngine` API, but according to the [Javadoc](https://junit.org/junit5/docs/current/api/org.junit.platform.engine/org/junit/platform/engine/TestEngine.html#getArtifactId()):

>Get the Artifact ID of the JAR in which this test engine is packaged. 

That's not the artifact ID? And why do we need a separate engine, for testing?

Reply:
It's because the one used by JUnit to test their display name generators is internal . In order to test this feature,  I get inspired by what they did

https://github.com/junit-team/junit5/blob/main/jupiter-tests/src/test/java/org/junit/jupiter/api/DisplayNameGenerationTests.java

Reply:
This doc is talking about the default implementation .
This article helped me a lot

https://blogs.oracle.com/javamagazine/post/junit-build-custom-test-engines-java

Reply:
> It's because the one used by JUnit to test their display name generators is internal . In order to test this feature, I get inspired by what they did
> 
> https://github.com/junit-team/junit5/blob/main/jupiter-tests/src/test/java/org/junit/jupiter/api/DisplayNameGenerationTests.java

I tried to extend the class **AbstractJupiterTestEngineTests** but it was not possible,  that's why I concluded that I have to implement my own Test Engine.
It was my first time to deal with Test engine.  While figuring out how to do that, I got this article from Oracle

https://blogs.oracle.com/javamagazine/post/junit-build-custom-test-engines-java

Reply:
Ah, I see. Thanks for the great explanation and those helpful links! 👍